src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage.cs

```diff
@@ -118,35 +118,31 @@ internal partial class SQLitePersistentStorage : AbstractPersistentStorage
 
         // cached query strings
 
-        private readonly string _select_star_from_string_table;
-        private readonly string _insert_into_string_table_values_0;
-        private readonly string _select_star_from_string_table_where_0_limit_one;
+        private readonly string _select_star_from_string_table = $@"select * from {StringInfoTableName}";
+        private readonly string _insert_into_string_table_values_0 = $@"insert into {StringInfoTableName}(""{DataColumnName}"") values (?)";
+        private readonly string _select_star_from_string_table_where_0_limit_one = $@"select * from {StringInfoTableName} where (""{DataColumnName}"" = ?) limit 1";
 
-        // We pool connections to the DB so that we don't have to take the hit of 
-        // reconnecting.  The connections also cache the prepared statements used
-        // to get/set data from the db.  A connection is safe to use by one thread
-        // at a time, but is not safe for simultaneous use by multiple threads.
-        private readonly object _connectionGate = new();
-        private readonly Stack<SqlConnection> _connectionsPool = new();
-
-        public SQLitePersistentStorage(
+        private SQLitePersistentStorage(
+            SQLiteConnectionPoolService connectionPoolService,
+            Solution? bulkLoadSnapshot,
             string workingFolderPath,
             string solutionFilePath,
             string databaseFile,
-            IDisposable dbOwnershipLock,
             IPersistentStorageFaultInjector? faultInjector)
             : base(workingFolderPath, solutionFilePath, databaseFile)
         {
-            _dbOwnershipLock = dbOwnershipLock;
-            _faultInjector = faultInjector;
+            _connectionPoolService = connectionPoolService;
 
             _solutionAccessor = new SolutionAccessor(this);
             _projectAccessor = new ProjectAccessor(this);
             _documentAccessor = new DocumentAccessor(this);
 
-            _select_star_from_string_table = $@"select * from {StringInfoTableName}";
-            _insert_into_string_table_values_0 = $@"insert into {StringInfoTableName}(""{DataColumnName}"") values (?)";
-            _select_star_from_string_table_where_0_limit_one = $@"select * from {StringInfoTableName} where (""{DataColumnName}"" = ?) limit 1";
+            _connectionPool = connectionPoolService.TryOpenDatabase(
+                bulkLoadSnapshot,
+                databaseFile,
+                faultInjector,
+                (bulkLoadSnapshot, connection, cancellationToken) => Initialize(bulkLoadSnapshot, connection, cancellationToken),
+                CancellationToken.None)!;
```

Comment:
trying to figure out hte `!` reasoning here.

Reply:
ah, i think i get it from seeing later code.  would be worthwhile to doc this.  i.e. that we do this to temporarily violate NRT rules, but that this then checked immediately after in TryCreate.

Reply:
:arrow_right: Now documented